File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/security/acl/AccessControlEntry.ts
module api.security.acl {

    import ArrayHelper = api.util.ArrayHelper;
    import Principal = api.security.Principal;

    export class AccessControlEntry implements api.Equitable {

        private static ALL_PERMISSIONS: Permission[] = [Permission.READ, Permission.CREATE, Permission.MODIFY, Permission.DELETE,
            Permission.PUBLISH,
            Permission.READ_PERMISSIONS, Permission.WRITE_PERMISSIONS
        ];

        private principal: Principal;

        private allowedPermissions: Permission[];

        private deniedPermissions: Permission[];

        private inherited: boolean;

        constructor(principal: Principal, inherited: boolean = false) {
            this.principal = principal;
            this.inherited = inherited;
            this.allowedPermissions = [];
            this.deniedPermissions = [];
        }

        getPrincipal(): Principal {
            return this.principal;
        }

        getPrincipalKey(): PrincipalKey {
            return this.principal.getKey();
        }

        getPrincipalDisplayName(): string {
            return this.principal.getDisplayName();
        }

        getAllowedPermissions(): Permission[] {
            return this.allowedPermissions;
        }

        getDeniedPermissions(): Permission[] {
            return this.deniedPermissions;
        }

        setAllowedPermissions(permissions: Permission[]): void {
            this.allowedPermissions = permissions;
        }

        setDeniedPermissions(permissions: Permission[]): void {
            this.deniedPermissions = permissions;
        }

        setInherited(inherited: boolean) {
            this.inherited = inherited;
        }

        isInherited(): boolean {
            return this.inherited;
        }

        isAllowed(permission: Permission): boolean {
            return (this.allowedPermissions.indexOf(permission) > -1) && (this.deniedPermissions.indexOf(permission) === -1);
        }

        isDenied(permission: Permission): boolean {
            return !this.isAllowed(permission);
        }

        isSet(permission: Permission): boolean {
            return (this.allowedPermissions.indexOf(permission) > -1) || (this.deniedPermissions.indexOf(permission) > -1);
        }

        allow(permission: Permission): AccessControlEntry {
            ArrayHelper.addUnique(permission, this.allowedPermissions);
            ArrayHelper.removeValue(permission, this.deniedPermissions);
            return this;
        }

        deny(permission: Permission): AccessControlEntry {
            ArrayHelper.addUnique(permission, this.deniedPermissions);
            ArrayHelper.removeValue(permission, this.allowedPermissions);
            return this;
        }

        remove(permission: Permission): AccessControlEntry {
            ArrayHelper.removeValue(permission, this.allowedPermissions);
            ArrayHelper.removeValue(permission, this.deniedPermissions);
            return this;
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, AccessControlEntry)) {
                return false;
            }

            var other = <AccessControlEntry>o;
            return this.principal.getKey().equals(other.getPrincipalKey()) &&
                   (this.principal.getDisplayName() == other.getPrincipalDisplayName()) &&
                   this.permissionEquals(this.allowedPermissions, other.allowedPermissions) &&
                   this.permissionEquals(this.deniedPermissions, other.deniedPermissions);
        }

        toString(): string {
            var values = '';
            AccessControlEntry.ALL_PERMISSIONS.forEach((permission: Permission) => {
                if (this.isSet(permission)) {
                    if (values !== '') {
                        values += ', ';
                    }
                    values += this.isAllowed(permission) ? '+' : '-';
                    values += Permission[permission].toUpperCase();
                }
            });
            return this.getPrincipalKey().toString() + '[' + values + ']';
        }

        clone(): AccessControlEntry {
            var ace = new AccessControlEntry(this.principal.clone());
            ace.allowedPermissions = this.allowedPermissions.slice(0);
            ace.deniedPermissions = this.deniedPermissions.slice(0);
            ace.inherited = this.inherited;
            return ace;
        }

        private permissionEquals(listA: Permission[], listB: Permission[]): boolean {
            return (listA.length === listB.length) &&
                   listA.every(function (element, idx) {
                       return element === listB[idx];
                   });
        }

        toJson(): api.security.acl.AccessControlEntryJson {
            return {
                "principal": this.principal.toJson(),
                "allow": this.allowedPermissions.map((perm) => Permission[perm]),
                "deny": this.deniedPermissions.map((perm) => Permission[perm])
            };
        }

        static fromJson(json: api.security.acl.AccessControlEntryJson): AccessControlEntry {
            var ace = new AccessControlEntry(Principal.fromJson(json.principal));
            var allow: Permission[] = json.allow.map((permStr) => Permission[permStr.toUpperCase()]);
            var deny: Permission[] = json.deny.map((permStr) => Permission[permStr.toUpperCase()]);
            ace.setAllowedPermissions(allow);
            ace.setDeniedPermissions(deny);
            return ace;
        }
    }

}

File: modules/wem-api/src/main/java/com/enonic/wem/api/security/acl/AccessControlEntry.java
package com.enonic.wem.api.security.acl;

import java.util.EnumSet;
import java.util.Objects;

import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Sets;

import com.enonic.wem.api.security.PrincipalKey;

import static com.google.common.base.Preconditions.checkNotNull;

public final class AccessControlEntry
{
    private final PrincipalKey principal;

    private final boolean inherited;

    private final ImmutableSet<Permission> allowedPermissions;

    private final ImmutableSet<Permission> deniedPermissions;

    private AccessControlEntry( final Builder builder )
    {
        this.principal = checkNotNull( builder.principal, "ACE principal cannot be null" );
        this.inherited = builder.inherited;
        this.allowedPermissions = Sets.immutableEnumSet( builder.allowedPermissions );
        this.deniedPermissions = Sets.immutableEnumSet( builder.deniedPermissions );
    }

    public PrincipalKey getPrincipal()
    {
        return principal;
    }

    public boolean isInherited()
    {
        return this.inherited;
    }

    public Iterable<Permission> getAllowedPermissions()
    {
        return allowedPermissions;
    }

    public Iterable<Permission> getDeniedPermissions()
    {
        return deniedPermissions;
    }

    public boolean isAllowed( final Permission permission )
    {
        return this.allowedPermissions.contains( permission ) && !this.deniedPermissions.contains( permission );
    }

    public boolean isDenied( final Permission permission )
    {
        return !this.isAllowed( permission );
    }

    public boolean isSet( final Permission permission )
    {
        return this.allowedPermissions.contains( permission ) || this.deniedPermissions.contains( permission );
    }

    @Override
    public String toString()
    {
        final StringBuilder sb = new StringBuilder( principal.toString() );
        sb.append( "[" );
        boolean empty = true;
        for ( Permission permission : Permission.values() )
        {
            if ( allowedPermissions.contains( permission ) )
            {
                if ( !empty )
                {
                    sb.append( ", " );
                }
                sb.append( "+" ).append( permission.toString().toLowerCase() );
                empty = false;
            }
            else if ( deniedPermissions.contains( permission ) )
            {
                if ( !empty )
                {
                    sb.append( ", " );
                }
                sb.append( "-" ).append( permission.toString().toLowerCase() );
                empty = false;
            }
        }
        sb.append( "]" );
        return sb.toString();
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( !( o instanceof AccessControlEntry ) )
        {
            return false;
        }
        final AccessControlEntry that = (AccessControlEntry) o;

        if ( !principal.equals( that.principal ) )
        {
            return false;
        }
        if ( !allowedPermissions.equals( that.allowedPermissions ) )
        {
            return false;
        }
        if ( !deniedPermissions.equals( that.deniedPermissions ) )
        {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( principal, allowedPermissions, deniedPermissions );
    }

    public static Builder create()
    {
        return new Builder();
    }

    public static Builder newACE( final AccessControlEntry ace )
    {
        return new Builder( ace );
    }

    public static class Builder
    {
        private PrincipalKey principal;

        private boolean inherited;

        private final EnumSet<Permission> allowedPermissions;

        private final EnumSet<Permission> deniedPermissions;

        private Builder()
        {
            this.allowedPermissions = EnumSet.noneOf( Permission.class );
            this.deniedPermissions = EnumSet.noneOf( Permission.class );
        }

        private Builder( final AccessControlEntry ace )
        {
            this.principal = ace.principal;
            this.inherited = ace.inherited;
            this.allowedPermissions = EnumSet.copyOf( ace.allowedPermissions );
            this.deniedPermissions = EnumSet.copyOf( ace.deniedPermissions );
        }

        public Builder principal( final PrincipalKey principal )
        {
            this.principal = principal;
            return this;
        }

        public Builder allow( final Permission permission )
        {
            this.deniedPermissions.remove( permission );
            this.allowedPermissions.add( permission );
            return this;
        }

        public Builder deny( final Permission permission )
        {
            this.allowedPermissions.remove( permission );
            this.deniedPermissions.add( permission );
            return this;
        }

        public Builder remove( final Permission permission )
        {
            this.allowedPermissions.remove( permission );
            this.deniedPermissions.remove( permission );
            return this;
        }

        public Builder inherited( final boolean inherited )
        {
            this.inherited = inherited;
            return this;
        }

        public Builder allowAll()
        {
            this.allow( Permission.CREATE );
            this.allow( Permission.DELETE );
            this.allow( Permission.PUBLISH );
            this.allow( Permission.READ );
            this.allow( Permission.MODIFY );
            this.allow( Permission.READ_PERMISSIONS );
            this.allow( Permission.WRITE_PERMISSIONS );

            return this;
        }

        public AccessControlEntry build()
        {
            return new AccessControlEntry( this );
        }
    }

}
